modules.d/90lvm-systemd/module-setup.sh

```diff
@@ -0,0 +1,92 @@
+#!/bin/bash
+# This file is part of dracut.
+# SPDX-License-Identifier: GPL-2.0-or-later
+
+# Prerequisite check(s) for module.
+check() {
+
+    [[ $hostonly ]] || [[ $mount_needs ]] && {
+        for fs in "${host_fs_types[@]}"; do
+            [[ $fs == LVM*_member ]] && return 0
+        done
+        return 255
+    }
+
+    # If the binary(s) requirements are not fulfilled the module can't be installed.
+    require_binaries \
+        lvm \
+        "$udevdir"/pvscan-udev-initrd.sh \
+        || return 1
+
+    # Return 0 to include the module.
+    return 0
+
+}
+
+# Module dependency requirements.
+depends() {
+
+    # This module has external dependency on other module(s).
+    echo rootfs-block dm systemd-udevd
+    # Return 0 to include the dependent module(s) in the initramfs.
+    return 0
+
+}
+
+# Commandline.
+cmdline() {
+    local _activated
+    declare -A _activated
+
+    for dev in "${!host_fs_types[@]}"; do
+        [ -e /sys/block/"${dev#/dev/}"/dm/name ] || continue
+        [ -e /sys/block/"${dev#/dev/}"/dm/uuid ] || continue
+        uuid=$(< /sys/block/"${dev#/dev/}"/dm/uuid)
+        [[ ${uuid#LVM-} == "$uuid" ]] && continue
+        dev=$(< /sys/block/"${dev#/dev/}"/dm/name)
+        eval "$(dmsetup splitname --nameprefixes --noheadings --rows "$dev" 2> /dev/null)"
+        [[ ${DM_VG_NAME} ]] && [[ ${DM_LV_NAME} ]] || return 1
+        if ! [[ ${_activated[DM_VG_NAME / DM_LV_NAME]} ]]; then
+            printf " rd.lvm.lv=%s " "${DM_VG_NAME}/${DM_LV_NAME} "
+            _activated["${DM_VG_NAME}/${DM_LV_NAME}"]=1
+        fi
+    done
+}
+
+# Install kernel module(s).
+installkernel() {
+    hostonly='' instmods dm-snapshot
+}
+
+# Install the required file(s) and directories for the module in the initramfs.
+install() {
+
+    if [[ $hostonly_cmdline == "yes" ]]; then
+        # shellcheck disable=SC2155
+        local _lvmconf=$(cmdline)
+        [[ $_lvmconf ]] && printf "%s\n" "$_lvmconf" >> "${initdir}/etc/cmdline.d/90lvm.conf"
+    fi
+
+    inst_hook cmdline 30 "$moddir/parse-lvm.sh"
+
+    inst_multiple -o \
+        "$udevdir"/pvscan-udev-initrd.sh \
+        "$udevrulesdir"/11-dm-lvm.rules \
+        "$udevrulesdir"/64-lvm.rules \
+        cache_dump cache_restore cache_check cache_repair \
+        era_check era_dump era_invalidate era_restore lvm \
+        thin_dump thin_restore thin_check thin_repair
+
+    # Install required libraries.
+    _arch=${DRACUT_ARCH:-$(uname -m)}
+    inst_libdir_file {"tls/$_arch/",tls/,"$_arch/",}"libdevmapper-event-lvm*.so"
+
+    # Install the hosts local user configurations if enabled.
+    if [[ $hostonly ]]; then
+        inst_multiple -H -o \
+            /etc/lvm/lvm.conf \
+            /etc/lvm/lvmlocal.conf
+    fi
+
+    dracut_need_initqueue
```

Comment:
Does it really need initqueue?

Reply:
@lnykryn well that's what upstream is doing here [1] @teigland ? 
We kinda need the scope of the change and why ( I could not find any discussion on the mailinglist about this upstream change ) 

1. https://sourceware.org/git/?p=lvm2.git;a=blob;f=dracut/module-setup.sh;h=51d9cd33a21686c666694c4235d049b1af8bb530;hb=c53bb2b5ea5e5a985e32235515ad21fff62e3b5b

Reply:
> Does it really need initqueue?

Probably not, I'm largely unaware of what most of this code is doing.  I took the previous code, changed the bits related to lvm that I did understand, and didn't touch most of the rest related to dracut that I didn't understand.  

> We kinda need the scope of the change and why

What do you mean, could you be more specific?

Reply:
> > We kinda need the scope of the change and why
> 
> What do you mean, could you be more specific?

Just an explanation as to what usecase(s) are you solving and why? 
Your upstream commit msg explains what you are doing but not why you have to do it, why the old module no longer suffices etc. at least that part is somewhat unclear to me. 

If we have the bigger picture of the change we can maybe help to improve it ( and maybe just improve the existing module so we dont have to go the three module route as in a meta module + lvm old module and lvm new module ).
 
For example I would rather we had a good generator and we called systemd type units from udev ( which we can put various conditions in the type units themselves, and include the the type unit in the initrd. target  )  as opposed to calling `RUN+="/usr/bin/systemd-run --no-block --property DefaultDependencies=no` foo and try to come up with udev ruleset that does not conflict with each other. `+` we need to rename the 64-lvm.rules to something else so it wont get confused with the existing 64-lvm.rules since those files are not identical.

Reply:
> Just an explanation as to what usecase(s) are you solving and why?

Activating the LVs listed by rd.lvm.lv (or VGs listed by rd.lvm.vg which is less common.)

> Your upstream commit msg explains what you are doing but not why you have to do it, why the old module no longer suffices etc. at least that part is somewhat unclear to me.

The old module does suffice, but it is inefficient and prone to failure with many devices.  It attempts to activate LVs after every lvm device appears rather than waiting for all the necessary devices to appear and then activating LVs.   lvm now knows when all the necessary devices are present, and we can use that intelligence to only run activation when it will succeed.

Repeated trying and failing to activate LVs (when there are multiple devices) is inefficient because each of those commands scans all devices.  As the number of devices grows, the number of failing activation commands becomes excessive, disk scanning becomes excessive and slow, and all the unnecessary overhead can lead to a boot timeout (emergency mode).

The new lvm capabilities allow us to scan each lvm device once, and then usually run a single activation command, which will then scan the disks a second time.

> For example I would rather we had a good generator and we called systemd type units from udev ( which we can put various conditions in the type units themselves, and include the type unit in the initrd. target ) as opposed to calling `RUN+="/usr/bin/systemd-run --no-block --property DefaultDependencies=no` foo 

I'm not a systemd expert so I don't know what advantages some other method might have.